include/revng/ADT/RecursiveCoroutine-coroutine.h

```diff
@@ -66,190 +69,204 @@ struct RecursivePromise : public ReturnBase<ReturnT> {
+    // suspend_always, because in the case that this coroutine was awaited by
+    // another RecursiveCoroutine, that awaiting coroutine must be resumed via
+    // symmetric transfer from this suspension point. To that end, a special
+    // awaiter is returned from this function.
     //
-    // So our final awaitable object should:
-    // - always suspend (await_ready should always return false)
-    // - return the handle to the awaiter continuation, so that if it's a
-    //   valid handle it will be resumed.
-    // - never be resumed, because it's the final suspension point
-    //   (await_resume actually aborts)
-    struct ResumeAwaiter {
-
-      ResumeAwaiter(std::coroutine_handle<> AwaiterHandle) :
-        Awaiter(AwaiterHandle) {}
+    // The final_suspend awaiter should:
+    // - Always suspend (await_ready returns false).
+    // - Resume the awaiting coroutine via symmetric transfer (await_suspend
+    //   returns the continuation handle (if any; otherwise noop_coroutine is
+    //   returned).
+    // - Never be resumed, because it's the final suspension point (await_resume
+    //   aborts).
+    struct Awaiter {
+
+      Awaiter(std::coroutine_handle<> Continuation) :
+        Continuation(Continuation) {}
 
       bool await_ready() const noexcept { return false; }
 
       std::coroutine_handle<> await_suspend(std::coroutine_handle<>) noexcept {
-        std::coroutine_handle<> ToResume = std::noop_coroutine();
-        if (Awaiter and not Awaiter.done()) {
-          ToResume = Awaiter;
-          Awaiter = {};
-        }
-        return ToResume;
+        return Continuation ? Continuation : std::noop_coroutine();
       }
 
       void await_resume() const noexcept { revng_abort(); }
 
     private:
-      std::coroutine_handle<> Awaiter;
+      std::coroutine_handle<> Continuation;
     };
 
-    auto ToResume = AwaiterContinuation;
-    AwaiterContinuation = {};
-    return ResumeAwaiter(ToResume);
+    // Transfers the continuation handle to the awaiter. The promise requires
+    // that no continuation is associated with it upon its destruction.
+    return Awaiter(std::exchange(Continuation, nullptr));
   }
 
-  RecursivePromise() : AwaiterContinuation(std::coroutine_handle<>{}) {}
-  ~RecursivePromise() { revng_assert(not AwaiterContinuation); }
+  RecursivePromise() : Continuation(nullptr) {}
+  ~RecursivePromise() { revng_assert(not Continuation); }
 
-  // Not copyable, otherwise AwaiterContinuation could be resumed twice.
+  // Not copyable, otherwise Continuation could be resumed twice.
   RecursivePromise &operator=(const RecursivePromise &) = delete;
   RecursivePromise(const RecursivePromise &) = delete;
 
-  RecursivePromise &operator=(RecursivePromise &&Other) {
-    if (this != &Other) {
-      this->AwaiterContinuation = Other.AwaiterContinuation;
-      Other.AwaiterContinuation = {};
-    }
-    return *this;
-  }
-  RecursivePromise(RecursivePromise &&Other) { *this = std::move(Other); }
-
```

Comment:
What happened to these?

I guess you're relying on the fact that the defaulted move constructor and assignment are generated by the compiler.
It's fine to use the defaulted move operators, but I want them explicitly defaulted. See [rule of 5](https://en.cppreference.com/w/cpp/language/rule_of_three.html)

Reply:
Oh nice. These don't need to be moveable.

Anyway, if they don't need to be moveable, then explicitly `delete` the move constructor and assignment.